impeller/renderer/backend/vulkan/fence_waiter_vk.cc

```diff
@@ -47,28 +48,25 @@ class WaitSetEntry {
 FenceWaiterVK::FenceWaiterVK(std::weak_ptr<DeviceHolder> device_holder)
     : device_holder_(std::move(device_holder)) {
   waiter_thread_ = std::make_unique<std::thread>([&]() { Main(); });
-  is_valid_ = true;
 }
 
 FenceWaiterVK::~FenceWaiterVK() {
   Terminate();
-  if (waiter_thread_) {
-    waiter_thread_->join();
-  }
-}
-
-bool FenceWaiterVK::IsValid() const {
-  return is_valid_;
+  waiter_thread_->join();
 }
 
 bool FenceWaiterVK::AddFence(vk::UniqueFence fence,
                              const fml::closure& callback) {
   TRACE_EVENT0("flutter", "FenceWaiterVK::AddFence");
-  if (!IsValid() || !fence || !callback) {
+  if (!fence || !callback) {
     return false;
   }
   {
+    // Maintain the invariant that terminate_ is accessed only under the lock.
```

Comment:
The locking without static analysis makes me nervous. I suspect we didn't use the `IPLR_GUARDED_BY` and friends because we don't have static thread safety analysis around condition variable. This should be easy enough to add support for. Can we file a followup bug for it?

Reply:
Sure, filed https://github.com/flutter/flutter/issues/134843.

Fwiw, this might be yet another place where absl fixes many of these problems for us by having this low-level infra support we don't have to maintain (https://github.com/flutter/flutter/issues/134741).